File: app/models/service.py
from __future__ import annotations

from datetime import datetime
from typing import List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class ServiceHostAssignment(BaseModel):
    host_id: int
    sort: Optional[int] = Field(default=None, ge=0)


class ServiceCreate(BaseModel):
    name: str = Field(..., max_length=128)
    description: Optional[str] = Field(None, max_length=256)
    hosts: List[ServiceHostAssignment] = Field(default_factory=list)
    admin_ids: List[int] = Field(default_factory=list)

    @field_validator("hosts")
    @classmethod
    def ensure_unique_hosts(cls, value: List[ServiceHostAssignment]):
        ids = [assignment.host_id for assignment in value]
        if len(ids) != len(set(ids)):
            raise ValueError("Duplicate host ids are not allowed in a service")
        return value


class ServiceModify(BaseModel):
    name: Optional[str] = Field(None, max_length=128)
    description: Optional[str] = Field(None, max_length=256)
    hosts: Optional[List[ServiceHostAssignment]] = None
    admin_ids: Optional[List[int]] = None

    @field_validator("hosts")
    @classmethod
    def ensure_unique_hosts(cls, value: Optional[List[ServiceHostAssignment]]):
        if value is None:
            return value
        ids = [assignment.host_id for assignment in value]
        if len(ids) != len(set(ids)):
            raise ValueError("Duplicate host ids are not allowed in a service")
        return value


class ServiceHost(BaseModel):
    id: int
    remark: str
    inbound_tag: str
    inbound_protocol: str
    sort: int
    address: str
    port: Optional[int]
    model_config = ConfigDict(from_attributes=True)


class ServiceAdmin(BaseModel):
    id: int
    username: str
    used_traffic: int
    lifetime_used_traffic: int
    model_config = ConfigDict(from_attributes=True)


class ServiceBase(BaseModel):
    id: int
    name: str
    description: Optional[str]
    used_traffic: int = 0
    lifetime_used_traffic: int = 0
    host_count: int = 0
    user_count: int = 0
    has_hosts: bool = True
    broken: bool = False
    model_config = ConfigDict(from_attributes=True)


class ServiceDetail(ServiceBase):
    admins: List[ServiceAdmin] = Field(default_factory=list)
    hosts: List[ServiceHost] = Field(default_factory=list)
    admin_ids: List[int] = Field(default_factory=list)
    host_ids: List[int] = Field(default_factory=list)


class ServiceListResponse(BaseModel):
    services: List[ServiceBase]
    total: int


class ServiceUsageResetResponse(BaseModel):
    id: int
    used_traffic: int
    lifetime_used_traffic: int
    model_config = ConfigDict(from_attributes=True)


class ServiceUsagePoint(BaseModel):
    timestamp: datetime
    used_traffic: int


class ServiceUsageTimeseries(BaseModel):
    service_id: int
    start: datetime
    end: datetime
    granularity: Literal["day", "hour"]
    points: List[ServiceUsagePoint]


class ServiceAdminUsage(BaseModel):
    admin_id: Optional[int]
    username: str
    used_traffic: int


class ServiceAdminUsageResponse(BaseModel):
    service_id: int
    start: datetime
    end: datetime
    admins: List[ServiceAdminUsage]


class ServiceAdminTimeseries(BaseModel):
    service_id: int
    admin_id: Optional[int]
    username: str
    start: datetime
    end: datetime
    granularity: Literal["day", "hour"]
    points: List[ServiceUsagePoint]


class ServiceDeletePayload(BaseModel):
    mode: Literal["delete_users", "transfer_users"] = "transfer_users"
    target_service_id: Optional[int] = None
    unlink_admins: bool = False

    @model_validator(mode="after")
    def validate_target(self):
        if self.target_service_id is not None and self.target_service_id < 1:
            raise ValueError("target_service_id must be a positive integer when provided")
        return self

File: dashboard/src/types/Service.ts
export type ServiceHost = {
	id: number;
	remark: string;
	inbound_tag: string;
	inbound_protocol: string;
	sort: number;
	address: string;
	port: number | null;
};

export type ServiceAdmin = {
	id: number;
	username: string;
	used_traffic: number;
	lifetime_used_traffic: number;
};

export type ServiceSummary = {
	id: number;
	name: string;
	description: string | null;
	flow: string | null;
	used_traffic: number;
	lifetime_used_traffic: number;
	host_count: number;
	user_count: number;
	has_hosts: boolean;
	broken: boolean;
};

export type ServiceDetail = ServiceSummary & {
	admins: ServiceAdmin[];
	hosts: ServiceHost[];
	admin_ids: number[];
	host_ids: number[];
};

export type ServiceListResponse = {
	services: ServiceSummary[];
	total: number;
};

export type ServiceHostAssignment = {
	host_id: number;
	sort?: number;
};

export type ServiceCreatePayload = {
	name: string;
	description?: string | null;
	flow?: string | null;
	admin_ids: number[];
	hosts: ServiceHostAssignment[];
};

export type ServiceModifyPayload = Partial<
	Omit<ServiceCreatePayload, "hosts">
> & {
	hosts?: ServiceHostAssignment[];
};

export type ServiceDeletePayload = {
	mode: "delete_users" | "transfer_users";
	target_service_id?: number | null;
	unlink_admins?: boolean;
};

export type ServiceUsagePoint = {
	timestamp: string;
	used_traffic: number;
};

export type ServiceUsageTimeseries = {
	service_id: number;
	start: string;
	end: string;
	granularity: "day" | "hour";
	points: ServiceUsagePoint[];
};

export type ServiceAdminUsage = {
	admin_id: number | null;
	username: string;
	used_traffic: number;
};

export type ServiceAdminUsageResponse = {
	service_id: number;
	start: string;
	end: string;
	admins: ServiceAdminUsage[];
};

export type ServiceAdminTimeseries = {
	service_id: number;
	admin_id: number | null;
	username: string;
	start: string;
	end: string;
	granularity: "day" | "hour";
	points: ServiceUsagePoint[];
};
